docs/source/tutorials/layer_memory_tracking.rst

```diff
@@ -0,0 +1,168 @@
+
+    suggestion = suggest_checkpoint_location(tracker.memory_traces, num_checkpoints=2)
+    print(suggestion.max_memory)      # Outputs: 485095936
+    print(suggestion.split_modules)   # Outputs: ['layer1.1.bn3', 'layer2.2.conv3']
+
+This sample code tells us that we can reduce the memory consumption due to activations from 1.4G to around 500M by
+checkpointing activations at the locations `layer1.1.bn3` and `layer2.2.conv3`.
+
+These locations can serve as first guesses and might not always be practical due to the model code. In the case of a
+torchvision resnet, we can adapt those locations by trying to checkpoint around layer1 and layer2:
+
+.. code-block:: python
+
+    model = torchvision.models.resnet50().cuda()
+    model.layer1 = checkpoint_wrapper(model.layer1)
+    model.layer3 = checkpoint_wrapper(torch.nn.Sequential(model.layer2, model.layer3))
+    model.layer2 = torch.nn.Identity()
+
+Leading to the following memory profile, saving around 400MB of activation memory at the cost of more compute:
+
+.. image:: _static/img/layer_memory_profile_optimized.png
+    :width: 500px
+    :align: center
+
+
+Dedicated features to FSDP distributed training
+-----------------------------------------------
+
+When training a big model with `FullyShardedDataParallem`, you can use the `LayerwiseMemoryTracker` to track the
+amount of memory exchanged by FSDP to consolidate sharded layers:
+
+.. code-block:: python
+
+    from fairscale.nn import FullyShardedDataParallel as FSDP
+    from fairscale.experimental.tooling.layer_memory_tracker import ProcessGroupTracker
+
+    # Create a process group for FSDP
+    group = torch.distributed.new_group()
+    group = ProcessGroupTracker(group)
+
+    # Create a FSDP model
+    model = torchvision.models.resnet50().cuda()
+    model.layer1 = FSDP(model.layer1, process_group=group)
+    model.layer2 = FSDP(model.layer2, process_group=group)
+    model.layer3 = FSDP(model.layer3, process_group=group)
+    model.layer4 = FSDP(model.layer4, process_group=group)
+    model = FSDP(model, process_group=group)
+
+Now, the `LayerwiseMemoryTracker` will provide an additional graph where we can see:
+
+- the memory spikes (in blue for forward, in orange for backward) of the `all_gather` calls
+- an estimation (in green) of cumulative parameter memory (only available for the forward pass)
+
+.. image:: _static/img/all_gathered_memory.png
+    :width: 500px
+    :align: center
+
+
+Limitations
+------------
+
+The `LayerwiseMemoryTracker` has a bunch of limitations it is important to be aware of:
+
+1. It only works on GPU models (models cannot sit on the CPU)
+2. Some of the GPU memory might not tracked by PyTorch (for example some NCCL buffers)
+3. Beside memory allocated and memory cached, which are based on PyTorch, the results are based on heuristic, and might miss some memory in some cases
+4. Some features (such as cumulatived all gathered memory for FSDP) do not work in the backward pass
```

Comment:
Do we know when this may not work? Is it based on the model and training setup?

Reply:
Yes, these features never work because of the asymmetry between forward and backward passes. For the forward, we get a "pre" and "post" hook, but we don't have this in the "backward" and so I could not implement it correctly.